handle GRUP in wellState from target

totto82 opened the pull request:
This seems to help significantly for some key models. 
The code for getting the well part of the group target is at least at a first glance ripe for refactoring. The similar logic is used 4 places. But since this fix is urgent, and the refactoring nontrivial, I suggest that happens another day. 

I have tested this on the norne family and on a proprietary model.

totto82 commented:
jenkins build this please

totto82 commented:
benchmark please

ytelses commented:
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1232

totto82 commented:
benchmark please

ytelses commented:
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 0.993    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 0.99     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 1.014    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.02     |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.762    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.772    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1233

totto82 commented:
jenkins build this please

atgeirr commented on opm/simulators/wells/WellInterfaceFluidSystem.cpp:
```diff
@@ -916,6 +918,181 @@ flowPhaseToEbosPhaseIdx(const int phaseIdx) const
+    if (currentGroupControl == Group::InjectionCMode::FLD ||
+        currentGroupControl == Group::InjectionCMode::NONE) {
+        if (!group.injectionGroupControlAvailable(injectionPhase)) {
+            return 1.0;
```
This, and also line 974, will give a target of 1 cubic meter per second, would it not be better to return an optional double, and set the target only if relevant? It looks like this may be taken from the production variant, which returns a scaling rather than a target (and a 1.0 scaling would be the same as doing nothing).

totto82 replied:
Good catch. It was a bit back-and-forth if the injector rates should be scaled or just set like we do now. Or will update as you suggest.

atgeirr reviewed (changes requested):
I think the logic is ok, I agree it should be refactored!

The 1.0 target should be fixed before merging.

totto82 commented on opm/simulators/wells/MultisegmentWell.hpp:
```diff
@@ -115,6 +115,7 @@ namespace Opm
 
         /// updating the well state based the current control mode
         void updateWellStateWithTarget(const Simulator& ebos_simulator,
```
Thanks for noticing See #3385

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

totto82 commented:
I have addressed the issues pointed out by @atgeirr  and fixed the critical test failures. The rest of the test failures are due to expected changes in the convergence and thus differences in the time-stepping. From my side this can be merged. I will created updated reference data later today if no objections.

totto82 commented:
jenkins build this update_data please

totto82 commented:
jenkins build this opm-tests=564 please

joakim-hove commented:
OK - we seem to have a data race going - **you go first**!

totto82 commented:
Ok, thanks.
